### DISCUSSION TITLE
React component library not showing

### DISCUSSION THREAD

**Discussion by saidinusah** [tailwindlabs/tailwindcss#11172] [Help]
Hello, I have my component library built with vite, but the main issue is the styling doesn't show.
Sometime it shows when i implement the same component. Eg. When i import the button from my component library it  doesn't show the styling but if i implement the same button with the same style and all, the imported component then shows with the styling. Please help

**Comment by wongjn** [comment]
Consider checking the `content` file globs sufficiently cover the component library files.

**Comment by rohvald** [comment]
1. Make sure that content in your tailwind.config.js includes all the files you need styled.

1. Add the lib config and components to the tailwind.config.js of the app. For some reason i had to use a relative path instead of an absolute for the presets: [ require('...') ]. My example:
```
/** @type {import('tailwindcss').Config} */
module.exports = {
  presets: [
    require('../../libs/moonlight-neon-ui/tailwind.config.js'),
  ],
  content: [
    'apps/ac-dcp/**/*.{html,js,jsx,tsx,ts}',
    'libs/moonlight-neon-ui/src/**/*.{html,js,jsx,tsx,ts}'
  ],
  theme: {
    extend: {}
  },
  plugins: [],
}
```

**Reply by AdsonCicilioti** [reply]
> 1. Make sure that content in your tailwind.config.js includes all the files you need styled.
> 2. Add the lib config and components to the tailwind.config.js of the app. For some reason i had to use a relative path instead of an absolute for the presets: [ require('...') ]. My example:
> 
> ```
> /** @type {import('tailwindcss').Config} */
> module.exports = {
>   presets: [
>     require('../../libs/moonlight-neon-ui/tailwind.config.js'),
>   ],
>   content: [
>     'apps/ac-dcp/**/*.{html,js,jsx,tsx,ts}',
>     'libs/moonlight-neon-ui/src/**/*.{html,js,jsx,tsx,ts}'
>   ],
>   theme: {
>     extend: {}
>   },
>   plugins: [],
> }
> ```

Nice tip!

Also i have this question. Certainly this work, but all styles from all components will be included on final css. The TailwindCss utils has any function to refine which components are be used on app fir built a final css more clean?

Believe that must use a logic function scanning the files.

**Reply by wongjn** [reply]
You could refine the `content` file globs to only the files you use.

**Reply by AdsonCicilioti** [reply]
> You could refine the `content` file globs to only the files you use.

Ok @wongjn, but imagine components that have properties that define a variant of colors, size or rounding, where not necessarily the user has used all variations.  If I just pass the used components path, I'll still be loading css from many classes I didn't use.  See the following file in my lib that defines the properties that my button components will have:
```typescript
import { cva, cx, VariantProps } from "class-variance-authority";

const buttonStyle = cva(
  cx([
    "inline-flex justify-center self-center items-center font-semibold",
    "active:scale-95 rounded cursor-pointer text-sm ease-in-out transition",
    "text-white",
  ]),
  {
    variants: {
      intent: {
        default: "bg-system-600 hover:bg-system-500 active:bg-system-700",
        primary: "bg-primary-600 hover:bg-primary-500 active:bg-primary-700",
        secondary:
          "bg-secondary-600 hover:bg-secondary-500 active:bg-secondary-700",
        accent: "bg-accent-600 hover:bg-accent-500 active:bg-accent-700",
        warning: "bg-amber-600 hover:bg-amber-500 active:bg-amber-700",
        danger: "bg-error-600 hover:bg-error-500 active:bg-error-700",
        success: "bg-success-600 hover:bg-success-500 active:bg-success-700",
      },
      size: {
        sm: "py-1 px-2 min-h-[1.75rem]  text-sm",
        md: "py-2 px-4 min-h-[2.5rem]",
        lg: "py-3 px-6 min-h-[3rem] text-lg",
        xl: "py-4 px-8 min-h-[3.75rem] text-xl",
      },

      roundness: {
        none: "rounded-none",
        sm: "rounded-sm",
        md: "rounded-md",
        lg: "rounded-lg",
        pill: "rounded-full",
      },
    },
    // compoundVariants: {},
    defaultVariants: {
      intent: "default",
      size: "md",
      roundness: undefined,
    },
  }
);

type ButtonProps = VariantProps<typeof buttonStyle>;

export { buttonStyle, type ButtonProps };
```

This is my React component that uses property definitions:
```typescript
import { Slot } from "@radix-ui/react-slot";

import { ButtonProps, buttonStyle, cn } from "@arakin-ui/core";

type ButtonType = React.ButtonHTMLAttributes<HTMLButtonElement> &
  ButtonProps & {
    children?: React.ReactNode;
    asChild?: boolean;
  };

export const Button: React.FC<ButtonType> = (props) => {
  const { children, asChild, intent, size, roundness, className, ...rest } =
    props;
  const ButtonComp = asChild ? Slot : "button";

  return (
    <>
      <ButtonComp
        className={cn(buttonStyle({ intent, size, roundness, className }))}
        {...rest}
      >
        {children}
      </ButtonComp>
    </>
  );
};

```

(I created the `@my-ui-lib/core` and `@my-ui/react` libs separately to be framework agnostic)

 So I can use my components like this:

```tsx
<div className="flex p-8 gap-6">
      <Button>Default</Button>
      <Button intent="primary" size={"lg"}>
          Primary
      </Button>
      <Button intent="secondary" roundness={"pill"} size={"sm"}>
          Secondary
      </Button>
</div>
```
Result:
![image](https://github.com/tailwindlabs/tailwindcss/assets/6278398/5c551d2a-25cb-46ad-a6d6-e804d1cbea67)

Imagine that, in a certain application, I use only these variants and props.  Everything else would be loaded.
 Now imagine this happening in other components.

The question is that some developers are looking for a way to take full advantage of the way TailwindCSS removes unused css, which in my opinion is awesome and is one of the strengths that makes me use it.

 At the moment I'm working on a script that uses `node:fs` (or just `fs`) that goes through all the application files, checking how the components were used and which properties that load TW classes were used and gives me a big string containing the classes.
 Then I include it in the `tailwind.config.js` file, in the `content` section as `raw`:

```javascript
require { getUsedComponentStyles } = require("@my-ui/core")

const usedComponentStyles = getUsedComponentStyles( scanDir: "./src", fileTypes: [".tsx", ".jsx"],  )

module.exports = {
  presets: [require("../../ui/core/tailwind.config.js")], 
  content: [
    "./index.html",
    "./src/**/*.{js,ts,jsx,tsx}",
    { raw: usedComponentStyles },
  ],
... // other configs
};
```

It works great, however in development it doesn't reload the `tailwind.config.js` file and the `getUSedComponentsStyles` function that traverses and finds the changes.  That's why I'm only using it in the production build.

It will be cool to see this working in development as well.  I'm still studying a way (I know I shouldn't worry because the important thing is that it works well in production).

 I scoured the tailwind source code to better understand how taiwind goes through the files and identifies its classes, I saw that they use Rust and I even think that it is responsible for this task.  I understood very little of what I saw and that's why I decided to create something in JS.
